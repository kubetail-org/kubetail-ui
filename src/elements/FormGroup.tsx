// Copyright 2024 Andres Morey
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { createContext, forwardRef, useMemo } from 'react';

import { cn } from '@/lib/utils';

type ContextState = {
  controlId: string | undefined,
};

const Context = createContext({} as ContextState); // for passing `controlId`
const baseCls = 'space-y-1';

interface Props extends React.ComponentPropsWithoutRef<'div'> {
  controlId?: string;
}

const FormGroup = forwardRef((
  {
    className,
    controlId,
    ...props
  }: Props,
  ref: React.ForwardedRef<HTMLDivElement>,
) => {
  const context = useMemo(() => ({ controlId }), [controlId]);

  return (
    <Context.Provider value={context}>
      <div
        {...props}
        ref={ref}
        className={cn(baseCls, className)}
      />
    </Context.Provider>
  );
});

FormGroup.displayName = 'FormGroup';

export { Context };
export default FormGroup;
