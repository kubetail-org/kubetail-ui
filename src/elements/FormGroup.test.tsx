// Copyright 2024-2025 Andres Morey
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { render } from '@testing-library/react';

import FormGroup from './FormGroup';

describe('FormGroup', () => {
  it('renders contents properly', () => {
    const { asFragment } = render(<FormGroup>contents</FormGroup>);
    expect(asFragment()).toMatchSnapshot();
  });

  it('appends custom className', () => {
    const { asFragment } = render(<FormGroup className="extra-class" />);
    expect(asFragment()).toMatchSnapshot();
  });
});
