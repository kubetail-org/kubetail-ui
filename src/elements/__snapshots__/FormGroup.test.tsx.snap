// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FormGroup > appends custom className 1`] = `
<DocumentFragment>
  <div
    class="space-y-1 extra-class"
  />
</DocumentFragment>
`;

exports[`FormGroup > renders contents properly 1`] = `
<DocumentFragment>
  <div
    class="space-y-1"
  >
    contents
  </div>
</DocumentFragment>
`;
